Add ability to redirect all prints

ShaharHD opened the pull request:
Added `output` (which defaults to `std::cout`) and `error` (which defaults to `std::cerr`) to allow easier redirect of all prints.

Useful when used in a daemon for example or when using a more complex system like [spdlog](https://github.com/gabime/spdlog)

ShaharHD committed:
Add std::ostream to all classes

ShaharHD committed:
redirect all output

FlorianReimold commented:
This looks really good. Awesome technique to just pass a stream. I always thought about the best technique for redirectable log output and this may be it.
I will take a deeper look after the holidays and merge this PR 👍

FlorianReimold commented on fineftp-server/include/fineftp/server.h:
```diff
@@ -52,8 +53,10 @@ namespace fineftp
+     * @param error: Error output prints. Defaults to std::cerr.
      */
-    FINEFTP_EXPORT FtpServer(const std::string& address, uint16_t port = 21);
+    FINEFTP_EXPORT FtpServer(const std::string& address, const uint16_t port = 21, std::ostream& output = std::cout, std::ostream& error = std::cerr);
```
`port` should not be const, as it is passed by value (will give warnings otherwise)

ShaharHD replied:
I have no issue reverting it, but I never saw an issue with a const for any C or C++ int types when by-value is used (by-ref might be an issue)
(I try to use `const` on every param method to avoid typos or incorrect assignment)

ShaharHD replied:
Can you give an example of where you might think it will fail?

ShaharHD commented:
> This looks really good. Awesome technique to just pass a stream. I always thought about the best technique for redirectable log output and this may be it. I will take a deeper look after the holidays and merge this PR 👍

Thank you :)  but you did 99.9% of the work with this awesome library 🥇 

I see it it failing on some windows builds in the CI/CD and pushed a fix - please trigger the workflow again.

ShaharHD committed:
Fix WIN32 typo

ShaharHD commented:
Also, can you please run `clang-tidy` and push?
I've tried to match your code style, but I cam more of a `clang-format` guy ...

ShaharHD committed:
Avoid unused-private-field warning

ShaharHD committed:
Add the missing option for the constructor (only address)

ShaharHD commented:
@FlorianReimold I've used CMake `FetchContent_Declare` with `PATCH_COMMAND` to test this in my code base - hence the commits to of fixing some additional issues.

Let me know if you prefer I squash those

Also, can you please run `clang-tidy` and push?
I've tried to match your code style, but I'm more of a `clang-format` guy ...

FlorianReimold committed:
Fixed clang tidy warnings

FlorianReimold commented:
I let clang tidy check the code. There were no additional warnings, except from the consts. I fixed them.

I am currently looking at the Constructor API. For instance, the following call is technically fine:

```cpp
FtpServer("127.0.0.1", my_stream);
```

However, this only sets the "normal" log stream, not the error stream, creating the following situation:
- normal messages are directed to `my_steam`
- error messages are directed to `std::cerr` by default

I don't really like that, I think any user who passes only 1 stream would expect both the normal output and error-output would end up in that stream. I suggest the following:

- Only passing 1 stream should make both outputs be redirected into that stream
_or_
- The API should force the user to also set the error stream, if they set the normal output stream

What do you think?

ShaharHD commented:
> I let clang tidy check the code. There were no additional warnings, except from the consts. I fixed them.
> 
> I am currently looking at the Constructor API. For instance, the following call is technically fine:
> 
> ```c++
> FtpServer("127.0.0.1", my_stream);
> ```
> 
> However, this only sets the "normal" log stream, not the error stream, creating the following situation:
> 
> * normal messages are directed to `my_steam`
> * error messages are directed to `std::cerr` by default
> 
> I don't really like that, I think any user who passes only 1 stream would expect both the normal output and error-output would end up in that stream. I suggest the following:
> 
> * Only passing 1 stream should make both outputs be redirected into that stream
>   _or_
> * The API should force the user to also set the error stream, if they set the normal output stream
> 
> What do you think?

Sounds like a good idea.

I can use `static_assert` to inforce if by default while still keeping a `#define` which can allow it.

ShaharHD committed:
Add constructors to inforce output and error

ShaharHD commented:
> I let clang tidy check the code. There were no additional warnings, except from the consts. I fixed them.
> 
Still curious as to why you got that error.
I'm using clang-17 and I see no error with that `const`

> I am currently looking at the Constructor API. For instance, the following call is technically fine:
> 
> ```c++
> FtpServer("127.0.0.1", my_stream);
> ```
> 
> However, this only sets the "normal" log stream, not the error stream, creating the following situation:
> 
> * normal messages are directed to `my_steam`
> * error messages are directed to `std::cerr` by default
> 
> I don't really like that, I think any user who passes only 1 stream would expect both the normal output and error-output would end up in that stream. I suggest the following:
> 
> * Only passing 1 stream should make both outputs be redirected into that stream
>   _or_
> * The API should force the user to also set the error stream, if they set the normal output stream
> 
> What do you think?

Added constructor methods to inforce both or none.

FlorianReimold committed:
Reduced constructor complexity and fixed some clang tidy warnings

FlorianReimold commented:
You didn't remove the default arguments on all constructors. So I did. But that further increased the number of constructors and I really started to hate the constructor complexity.

So I simply removed everything but the most powerful constructor (and the old constructors to provide API stability and easy usage). I think, that it's acceptable to make developers always provide an address and port, if they want to redirect log output. I am just still looking for a good solution of oppressing all output. I think C++ doesn't provide a NULL stream.

FlorianReimold committed:
Reverted to original whitespace layout

ShaharHD commented:
No issues from my end.
I used only the most complex constructure in my code.

I tried to do the "force non-null" with `static_assert`, but it was even more cluttered which is why I went with the constructor approach.

I'm happy with what ever solution you are feeling comfortable with.
